`Assessment`: Allow instructors to select csv export options before downloading course scores

maximiliansoelch opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
When course scores get downloaded, the field separator as well as the quote strings are currently hardcoded and can not be customized by the user.


### Description
<!-- Describe your changes in detail -->
When the user clicks on the export button for downloading course scores, a new modal view is opened through which a custom field separator as well as custom quote strings can be selected.

Currently the newly added components are only used when exporting course scores, but in general they can be used everywhere in the project where csv exports are possible.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Course
- 1 Student
- 1 Exercise, on which the student participated

1. Log in to Artemis
2. Navigate to Course Administration, select the course, then Scores 
3. Click the `Export as CSV` button
4. Select a field separator as well as a the quote string that should be used for generating the csv file
5. Click `Export` 
6. View the downloaded csv file in a text editor and make sure the previously selected options were applied 

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
| Class/File | Branch | Line |
|------------|-------:|-----:|
| csv-export-button.component.ts | 75% | 100% |
| csv-export-modal.component.ts | 85.71% | 96.42% |
| course-scores.component.ts | 73.84% | 94.72% |
<!--
| ExerciseService.java | 85% | 77% |
| programming-exercise.component.ts | 13% | 95% |
-->


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![csv-export](https://user-images.githubusercontent.com/9954674/160402405-b31fe394-41c5-459c-aff9-3154a8fc6fa1.gif)

b-fein reviewed (commented):
Works as expected when testing with various export options. Nice work, especially already preparing that the same settings can be easily re-used in the future for the other exports. :+1:

I left two suggestions in the code. I did not do a detailed code-review of the HTML changes, though, as others are more familiar with best practices for the layout.

+1 on the quoting of the column headers as already suggested. As this is handled by the CSV library, though, I’m not sure if that can be easily fixed.
⇒ If it can be implemented, then I’d like you to. If that is a limitation by the library, then I’d approve without changing that, as the remaining functionality works as intended.

maximiliansoelch commented on src/main/webapp/app/shared/export/csv-export-modal.component.html:
```diff
@@ -0,0 +1,94 @@
+    </div>
+    <div class="modal-footer justify-content-between">
+        <div class="flex-grow-1 d-flex justify-content-end">
+            <button type="button" class="btn btn-default cancel" data-dismiss="modal" (click)="cancel()">
```
Adapted

maximiliansoelch commented on src/main/webapp/app/shared/export/csv-export-modal.component.html:
```diff
@@ -0,0 +1,94 @@
+                        (click)="setCsvFieldSeparator(CsvFieldSeparator.COMMA)"
+                    >
+                        {{ 'export.csvFile.fieldSeparator.comma' | artemisTranslate }}
+                    </div>
```
Code changed to your suggestion.

maximiliansoelch commented:
@bassner Thanks for the comment. I actually thought the export library will take care of that and therefore missed that the header row is not quoted according to the user's specified setting. I adjusted it now so that the user requested quotestring is also applied to the header row.

@b-fein Thanks for the review and testing. I implemented the changes according to your suggestions regarding the space between the export and cancel button as well as now utilizing ngfor to generate the button groups.

valeriebucher reviewed (approved):
Export works as expected!

bassner reviewed (approved):
Thanks for adding the changes!
The CSV column titles are now properly quoted.
Selected options are applied correctly, and the export works correctly.

Code looks good to me as well 👍
